### PR TITLE
feat(operator): add skipWhile operator

### PR DESCRIPTION
**Pull request by jinroh**
(#696)

Waiting for your reviews !

### CODE REVIEW

File: spec/operators/skipWhile-spec.js [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+
+    expectObservable(source.skipWhile(function () { return true; })).toBe(expected);
+  });
+
+  it('should skip all elements with a truthy predicate', function () {
+    var source = hot('-1-^2--3--4--5--6--|');
+    var expected =      '----------------|';
+
+    expectObservable(source.skipWhile(function () { return {}; })).toBe(expected);
+  });
```
Comment (staltz) [inline spec/operators/skipWhile-spec.js]: Can you add a test case where the predicate throws an error?

File: spec/operators/skipWhile-spec.js [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+  it('should be possible to skip using the element\'s index', function () {
+    var source = hot('--a--b-^-c--d--e--f--g--h--|');
+    var expected =          '--------e--f--g--h--|';
+
+    var predicate = function (v, index) {
+      return index < 2;
+    };
+
+    expectObservable(source.skipWhile(predicate)).toBe(expected);
+  });
```
Comment (staltz) [inline spec/operators/skipWhile-spec.js]: Would be good to add subscription assertions for all these test cases.

File: spec/operators/skipWhile-spec.js (hunk 1 above)
Comment (jinroh) [inline spec/operators/skipWhile-spec.js]: Forgot this one !

File: spec/operators/skipWhile-spec.js (hunk 2 above)
Comment (jinroh) [inline spec/operators/skipWhile-spec.js]: Done.

**Comment by jinroh** [comment]
In RxJS4, the predicate function [takes as third argument](https://github.com/Reactive-Extensions/RxJS/blob/master/src/core/linq/observable/skipwhile.js#L29) (after the value and index) the observable object being subscribed...

I'm not sure what is the purpose of this and could not find any... Do you think we need to keep it ?

File: src/operators/skipWhile.ts [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+  private skipping: boolean = true;
+  private index: number = 0;
+
+  constructor(destination: Subscriber<T>,
+              private predicate: (x: T, index: number) => boolean) {
+    super(destination);
+  }
+
+  _next(value: T): void {
+    const destination = this.destination;
```
Comment (staltz) [inline src/operators/skipWhile.ts]: Like for everything else, you could store `const skipping = this.skipping` since it's used multiple times.

Comment (jinroh) [inline src/operators/skipWhile.ts]: Not really because it is mutated on line 40.

Comment (staltz) [inline src/operators/skipWhile.ts]: Hmm, correct.

File: src/operators/skipWhile.ts [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+import {tryCatch} from '../util/tryCatch';
+import {errorObject} from '../util/errorObject';
+import {bindCallback} from '../util/bindCallback';
+
+export function skipWhile<T>(predicate: (x: T, index: number) => boolean, thisArg?: any) {
+  return this.lift(new SkipWhileOperator(predicate, thisArg));
+}
+
+class SkipWhileOperator<T, R> implements Operator<T, R> {
+  predicate: (x: T, index: number) => boolean;
```
Comment (kwonoj) [inline src/operators/skipWhile.ts]: maybe this can be private? minor thing though.

**Comment by kwonoj** [comment]
I think PR's OK. @jinroh , you're fast! :)

**Comment by blesh** [comment]
> I'm not sure what is the purpose of this and could not find any... Do you think we need to keep it ?

I believe we kept it in `filter` and `map` as well.. so we should probably keep it here too... however, I totally agree that it's unnecessary.

File: src/operators/skipWhile.ts [hunk 5]
```diff
@@ -0,0 +1,47 @@
+import {Operator} from '../Operator';
+import {Subscriber} from '../Subscriber';
+import {tryCatch} from '../util/tryCatch';
+import {errorObject} from '../util/errorObject';
+import {bindCallback} from '../util/bindCallback';
+
+export function skipWhile<T>(predicate: (x: T, index: number) => boolean, thisArg?: any) {
```
Comment (blesh) [inline src/operators/skipWhile.ts]: This should define a return value of `: Observable<T>`

**Comment by jinroh** [comment]
@blesh I don't think [you](https://github.com/ReactiveX/RxJS/blob/master/src/operators/filter.ts#L44) [did](https://github.com/ReactiveX/RxJS/blob/master/src/operators/map.ts#L44).

I think we should drop this behaviour...

**Comment by kwonoj** [comment]
Actually, there are mixture currently, refer [`count`](https://github.com/ReactiveX/RxJS/blob/master/src/operators/count.ts#L63). Seems it's good time to have conclusion for unified behavior.

File: src/operators/skipWhile.ts (hunk 5 above)
Comment (jinroh) [inline src/operators/skipWhile.ts]: done.

**Comment by blesh** [comment]
Oops didn't mean to close. Fat fingers.

Anyhow...

> I think we should drop this behaviour

That's fine. It really doesn't make much sense anyhow.

**Comment by kwonoj** [comment]
> That's fine. It really doesn't make much sense anyhow.

: +1 too. Maybe can discuss later to find if anyone has usecase.

**Comment by kwonoj** [comment]
I think most of issues in this PR's now settled down, will merge in around hr. (having around hr of buffer just in case if @blesh might have some additional suggestion)

**Comment by blesh** [comment]
@kwonoj :+1: LGTM

**Comment by kwonoj** [comment]
Merged with a2244e0. Appreciate for all effort & taking time, @jinroh

**Pull request closed without merging**